src/output.cpp

```diff
@@ -1081,6 +1271,93 @@ void JsonOutput::map_hist(
   out_ << "}}" << std::endl;
 }
 
+void JsonOutput::map_tseries(
+    BPFtrace &bpftrace,
+    const BpfMap &map,
+    const TSeriesMap &values_by_key,
+    const std::vector<std::pair<KeyType, EpochType>> &latest_epoch_by_key) const
+{
+  const auto &map_info = bpftrace.resources.maps_info.at(map.name());
+
+  out_ << R"({"type": ")" << MessageType::tseries << R"(", "data": {)";
+  out_ << "\"" << json_escape(map.name()) << "\": ";
+  if (!map_info.is_scalar) // check if this map has keys
+    out_ << "{";
+
+  map_tseries_contents(bpftrace, map, values_by_key, latest_epoch_by_key);
+
+  if (!map_info.is_scalar) // check if this map has keys
+    out_ << "}";
+  out_ << "}}" << std::endl;
+}
+
+template <typename T>
+std::string tseries_to_str_json(BPFtrace &bpftrace,
+                                const TSeries &values,
+                                EpochType last_epoch,
+                                uint64_t interval_ns,
+                                uint64_t num_intervals)
+{
+  EpochType first_epoch = util::tseries_first_epoch(values,
+                                                    last_epoch,
+                                                    num_intervals);
+  std::ostringstream res;
+
+  if (!first_epoch) {
+    return "[]";
+  }
+
+  bool first = true;
+  res << "[";
+  for (EpochType epoch = first_epoch; epoch <= last_epoch; epoch++) {
+    const auto &v = values.find(epoch);
+    if (v == values.end()) {
+      continue;
+    }
+
+    if (!first) {
+      res << ",";
+    } else {
+      first = false;
+    }
+
+    T val = 0;
+    if (v != values.end()) {
```

Comment:
nit: don't think we need this check again here as we checked at the top of the for loop

Reply:
Oops, I'll clean this up. I think I missed it during my last refactor.